clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3149,6 +3145,62 @@ void Sema::CheckSYCLKernelCall(FunctionDecl *KernelFunc, SourceRange CallLoc,
     KernelFunc->setInvalidDecl();
 }
 
+// For a wrapped parallel_for, copy attributes from original
+// kernel to wrapped kernel.
+void Sema::copySYCLKernelAttrs(const CXXRecordDecl *KernelObj) {
+  // Get the operator() function of the wrapper
+  CXXMethodDecl *OpParens = nullptr;
+  for (auto *MD : KernelObj->methods()) {
+    if (MD->getOverloadedOperator() == OO_Call) {
+      OpParens = MD;
+      break;
+    }
+  }
+  assert(OpParens && "invalid kernel object");
+
+  typedef std::pair<FunctionDecl *, FunctionDecl *> ChildParentPair;
+  llvm::SmallPtrSet<FunctionDecl *, 16> Visited;
+  llvm::SmallVector<ChildParentPair, 16> WorkList;
+  WorkList.push_back({OpParens, nullptr});
+  FunctionDecl *KernelBody = nullptr;
+
+  CallGraph SYCLCG;
+  SYCLCG.addToCallGraph(getASTContext().getTranslationUnitDecl());
+  while (!WorkList.empty()) {
+    FunctionDecl *FD = WorkList.back().first;
+    FunctionDecl *ParentFD = WorkList.back().second;
+
+    if ((ParentFD == OpParens) && isSYCLKernelBodyFunction(FD)) {
+      KernelBody = FD;
+      break;
+    }
+
+    WorkList.pop_back();
+    if (!Visited.insert(FD).second)
+      continue; // We've already seen this Decl
+
+    CallGraphNode *N = SYCLCG.getNode(FD);
+    if (!N)
+      continue;
+
+    for (const CallGraphNode *CI : *N) {
+      if (auto *Callee = dyn_cast<FunctionDecl>(CI->getDecl())) {
+        Callee = Callee->getMostRecentDecl();
+        if (!Visited.count(Callee))
+          WorkList.push_back({Callee, FD});
+      }
+    }
+  }
+
+  assert(KernelBody && "improper parallel_for wrap");
+  if (KernelBody) {
+    llvm::SmallVector<Attr *, 4> Attrs;
+    collectSYCLAttributes(*this, KernelBody, Attrs);
+    if (!Attrs.empty())
+      OpParens->setAttrs(Attrs);
```

Comment:
This call will assert if `OpParens` already has attributes on it (otherwise, it would blow away the existing attributes silently). Should this be: `llvm::for_each(Attrs, [OpParens](Attr *A) { OpParens->addAttr(A); });` to ensure that can't happen.

Reply:
We wrapped kernel is generated by the SYCL runtime (through handler.hpp) and we know that is has no attributes. So it is ok to simply set the attrs.

Reply:
The frontend is free to add implicit attributes onto any declaration it wants, so "no attributes in source" doesn't necessarily imply "no attributes in the AST". If you prefer to keep the `setAttrs()` call despite that, please assert that the declaration has no existing attributes before making the call. (This will also help people reading the code later to understand that this code isn't actually a bug.)